Keyboard navigation between components

somebody1234 opened the pull request:
### Pull Request Description
- Close https://github.com/enso-org/cloud-v2/issues/982
  - Add keyboard navigation via arrows between different components
    - This is achieved by a `Navigator2D` class which keeps track of the closest adjacent elements.

Other changes:
- Switch much of the codebase to use `react-aria-components`
  - This *should* (but does not necessarily) give us improved accessibility for free.
- Refactor various common styles into styled components
  - `FocusArea` to perform automatic registration with `Navigator2D`
  - `Button` and `UnstyledButton` to let buttons participate in keyboard navigation
  - `HorizontalMenuBar` - used for buttons below the titles in the Drive page, Keyboard Shortcuts settings page, and Members List settings page
  - `SettingsPage` in the settings pages
  - `SettingsSection` in the settings page to wrap around `FocusArea` and the heading for each section
- Add debugging utilities
  - Add debugging when `body` has the `data-debug` attribute: `document.body.dataset.debug = ''`
    - This adds rings around elements (all with different colors):
      - That are `FocusArea`s. `FocusArea` is a wrapper component that makes an element participate in `Navigator2D`.
      - That are `:focus`ed, and that are `:focus-visible`
      - That are `.focus-child`. This is because keyboard navigation via arrows ***ignores*** all focusable elements that are not `.focus-child`.
  - Debug `Navigator2D` neighbors when `body` has the `debug-navigator2d` attribute: `document.body.dataset.debugNavigator2d = ''`
    - This highlights neighbors of the currently focused element. This is a separate debug option because computing neighbors is potentially quite expensive.

### Important Notes
- :warning: Modals and the authentication flow are not yet fully tested.
- Up+Down to navigate through suggestions has been disabled to improve UX when accidentally navigating upwards to the assets search bar.
- There are a number of *known* issues with keyboard navigation. For the most part it's because a proper solution will be quite difficult.
  - Focus is lost when a column (from the extra columns selector) is toggled - because the button stops existing
  - It's not possible to navigate to the icons on the assets table - so it's current not possible to *hide* columns via the keyboard
  - Neighbors of the extra columns selector are not ideal (both when it is being navigated from, and when it is being navigated to)
  - The suggestions in the `AssetSearchBar` aren't *quite* fully integrated with arrow keyboard navigation.
    - This is *semi*-intentional. I think it makes a lot more sense to integrate them in, *however* it stays like this for now largely because I think pressing `ArrowUp` then `ArrowDown` from the assets table should return to the assets table
  - Likewise for the assets table. The reason here, however, is because we want multi-select. While `react-aria-components` has lists which support multi-select, it doesn't allow programmatic focus control, making it not particularly ideal, as we want to focus the topmost element when navigating in from above.
  - Clicking on the "New Folder" icon (and the like) do not focus on the newly created child. This one should be pretty easy to do, but I'm not sure whether it's the right thing to do.

### Testing Instructions
- The main purpose of this PR is to make sure that keyboard navigation is now possible, on ***every*** page/screen.
  - This includes settings pages and modals.
  - I think it doesn't have to be *perfect* for now, although it is definitely a good idea to find places where UX isn't quite up to par. For example:

- However, a lot of the rest of the codebase has undergone some refactoring, so most other UI actions SHOULD probably still be tested.
- There should be no changes to the logic for interacting with the backend however, so backend interactions SHOULD NOT need to be tested (although it does not hurt to test a few of them).

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

somebody1234 committed:
Initial (untested) implementation of 2D navigation

somebody1234 committed:
Add `Navigator2D` provider; initial adding to elements that should be
navigable to

somebody1234 committed:
Adjust distance algorithm for `Navigator2D`

somebody1234 committed:
Begin adding `primaryChild`; begin adding keyboard selection for children of containers

Also fix bug in `Navigator2D` vertical distance

somebody1234 committed:
Keyboard navigation and focus ring for Category Switcher

somebody1234 committed:
Some fixes; enable navigation for Labels panel

somebody1234 committed:
Fix bug with bounding box; fix navigating into Category Switcher

somebody1234 committed:
Add keyboard navigation to page switcher

somebody1234 committed:
Add keyboard navigation to search bar

somebody1234 committed:
Add keyboard navigation to top right menus; make `.focus-ring:focus` override `*:focus`

somebody1234 committed:
Add focus ring support to User Menu entries

somebody1234 committed:
Remove unnecessary colors

somebody1234 committed:
Add focus ring support to Settings Sidebar entries; remove focus ring on click

somebody1234 committed:
Fix arrow keys being mysteriously eaten

somebody1234 committed:
Focus settings sidebar when settings is opened

somebody1234 committed:
WIP

somebody1234 committed:
Merge branch 'develop' into wip/sb/keyboard-navigate-between-sections

somebody1234 committed:
Add keyboard navigation to User Account settings page

Change Password navigates to sidebar when pressing down...

somebody1234 committed:
Add keyboard navigation to Organization settings page

Organization section navigates to page switcher instead of settings sidebar when pressing left...

somebody1234 committed:
Minor fixes

somebody1234 committed:
Add keyboard navigation to Members settings page

Members button bar cannot navigate right...

somebody1234 committed:
Minor fixes for Dropdown styles

somebody1234 committed:
WIP

somebody1234 committed:
Minor fixes to "delete keyboard shortcut" button

somebody1234 committed:
Add keyboard navigation to "Keyboard shortcuts" table

Pressing up/down sometimes breaks further up/down presses

somebody1234 committed:
Prevent default on keyboard navigation to prevent unwanted scroll when keyboard navigating

Appears to have also fixed the up/down arrows sporadically breaking on
Keyboard shortcuts table in settings

somebody1234 committed:
Minor refactor; fix 'Enter' no longer working

Position is sometimes lost when pressing a button in keyboard shortcuts table.
This is likely due to a re-render causing the clicked button to disappear.

somebody1234 committed:
Add keyboard navigation to Drive Bar

somebody1234 committed:
Add keyboard navigation to Samples list and news items on home page

somebody1234 committed:
Fixes for `AssetSearchBar` and `AssetsTable`

somebody1234 committed:
Fix selection always disappearing

somebody1234 committed:
Merge branch 'develop' into wip/sb/keyboard-navigate-between-sections

somebody1234 committed:
Minor fixes

MrFlashAccount commented:
The fact that we need to write so much code to make it work each time is sooo annoying. I think we should investigate the use cases and hide as much as possible behind proper abstrations. 

Also worth noticing: in the Diff PR I added [react-aria-components](https://react-spectrum.adobe.com/react-aria/components.html), they comes with proper keyboard navigation out of the box. Maybe you take a look at them?

somebody1234 commented:
do they include arrow key support though?

MrFlashAccount commented:
For proper components - yes, also they expose hooks and `KeyboardManager` to implement your ouw custom navigation within a component. (Navigation between different components would be better to implement using `Tab` navigation, as it's done in vscode, for example)

somebody1234 commented:
fair enough... react-aria-components looks nice, but i think it might be out of scope of this PR (since ideally we would want to switch all components to react-aria)

(however, it's true that a lot of this PR would be unnecessary if we use react-aria-components, mostly the navigation between children)

somebody1234 commented:
> Navigation between different components would be better to implement using `Tab` navigation

yes, that should already Just Work based on order in markup though.

plus, i'm not sure it's a reason to *avoid* allowing arrow keys for navigation

somebody1234 committed:
WIP: Migrate to react-aria-components (part 1)

somebody1234 committed:
WIP: Migrate to react-aria-components (part 2)

somebody1234 committed:
WIP: Migrate to react-aria-components (part 3)

somebody1234 committed:
Change `Navigator2D` to focus area nearest to the child,  rather than the entire container

somebody1234 committed:
Add refs back to settings inputs

somebody1234 committed:
Further fixes to `Navigator2D` algorithm

somebody1234 committed:
Fixes for `Navigator2D` interacting with text inputs

somebody1234 committed:
Finally fix focus issue

somebody1234 committed:
Simplify `Navigator2D`

somebody1234 committed:
Fix some focus areas not appearing

somebody1234 committed:
WIP: Migrate to react-aria-components (part 4)

somebody1234 committed:
WIP: Migrate to react-aria-components (part 5)

somebody1234 committed:
Merge branch 'develop' into wip/sb/keyboard-navigate-between-sections-2

somebody1234 committed:
WIP: Migrate to react-aria-components (part 6)

somebody1234 committed:
WIP: Migrate to react-aria-components (part 7)

somebody1234 committed:
WIP: Migrate to react-aria-components (part 8)

somebody1234 committed:
WIP: Migrate to react-aria-components (part 9)

Next steps:

Fix focus issues in category switcher

Migrate WhatsNew/Samples to grids

somebody1234 committed:
Fix focus issues in `CategorySwitcher`

Navigating *back to* `CategorySwitcher` is still broken though.

somebody1234 committed:
Give up on `Menu`/`GridList`

somebody1234 committed:
Add debugging for `Navigator2D`; fix `Navigator2D` distance calculation

Arrow key navigation away from `AssetSearchBar` is still broken

somebody1234 committed:
Re-add focusing first child

somebody1234 committed:
Fix label selection

somebody1234 committed:
Fix settings sidebar focus

somebody1234 committed:
Fix most settings focus issues

Activity log bar is still problematic

somebody1234 committed:
Fix most other navigation focus issues

somebody1234 committed:
Fix focus issues for context menus and user menu

somebody1234 committed:
Skip over loading/empty rows when navigating `AssetsTable`

somebody1234 committed:
WIP: Keyboard navigation for `InviteUsersModal`

somebody1234 committed:
WIP: More keyboard navigation for modals

somebody1234 commented:
pinging @MrFlashAccount for opinions on general code style of this refactor

somebody1234 committed:
Add "focus roots" to `Navigator2D`

somebody1234 committed:
Fixes for keyboard navigation in modals; add focus ring to modals

somebody1234 committed:
Migrate modals to `ButtonRow`, `UnstyledButton`, and aria components

somebody1234 committed:
WIP: Add lints to enforce usage of React Aria Components; begin fix lints

somebody1234 committed:
Merge branch 'develop' into wip/sb/keyboard-navigate-between-sections

somebody1234 committed:
Merge branch 'develop' into wip/sb/keyboard-navigate-between-sections

somebody1234 committed:
Fix lints

somebody1234 committed:
Refactor some styles out of settings components

somebody1234 committed:
Add `HorizontalMenuBar`

somebody1234 committed:
Merge branch 'develop' into wip/sb/keyboard-navigate-between-sections

somebody1234 committed:
Fix React error

Also fix `/favicon.ico` warning in dev mode

somebody1234 committed:
Keyboard navigation improvements for `ColorPicker`

somebody1234 committed:
Immediately mark element as focused in `Navigator2D` if it is already focused

somebody1234 committed:
Fix vertical navigation in Keyboard Shortcuts table

somebody1234 committed:
Fix horizontal navigation in Keyboard Shortcuts table

somebody1234 committed:
Fix keyboard navigation for "New Secret" modal

somebody1234 committed:
Fix styling of (currently unused) checkbox

somebody1234 committed:
Adjust keyboard navigation for dropdowns

somebody1234 committed:
Finish arrow key navigation for Data Link modal

somebody1234 committed:
Add focus rings to `PermissionSelector` and `Autocomplete`

somebody1234 committed:
Add component for inputs that are focus children

somebody1234 committed:
Arrow key navigation for `ManagePermissionsModal`

somebody1234 committed:
Fix `FocusRoot`

somebody1234 committed:
Copy `RadioGroup` without the `keydown` handler to fix `ColorPicker`

somebody1234 committed:
Add keyboard navigation to `Login` page

somebody1234 committed:
Add keyboard navigation to the rest of the auth pages

somebody1234 committed:
Add autofocus to auth pages

somebody1234 committed:
Prettier

somebody1234 committed:
WIP: Fix some E2E tests; begin fixing asset name inputs

Next order of business is fixing search bar

somebody1234 committed:
Fix submittability of `EditableSpan`

somebody1234 committed:
Fix editing asset name? also refactor `useOnScroll` out

somebody1234 committed:
Fix `useOnScroll`

somebody1234 committed:
Fix "Change Password" form

somebody1234 committed:
Fix `AssetSearchBar`?

somebody1234 committed:
Fix tests involving `locateNameColumnHeading`

somebody1234 committed:
Fix E2E test dragging from the wrong place

somebody1234 committed:
FIx cancelling editing name

somebody1234 committed:
Fail at fixing "move to trash" E2E test

somebody1234 committed:
Fix "suggestions" E2E test

somebody1234 committed:
Fix labels

somebody1234 committed:
Fix final E2E test

somebody1234 committed:
Prettier

MrFlashAccount commented on app/ide-desktop/lib/dashboard/vite.config.ts:
```diff
@@ -45,3 +48,25 @@ export default vite.defineConfig({
 })
+
+/** A plugin to serve a favicon, in development mode only. */
+function serveFavicon(): vite.Plugin {
```
Why do we need to serve favicon manually instead of putting it in `./public` folder? (https://vitejs.dev/guide/assets#the-public-directory)

somebody1234 replied:
i assume the domain has a real favicon, this way there's only one source of truth on the actual deploy

MrFlashAccount replied:
I didn't get you, afaik we don't use vite to serve our app in production, right?

somebody1234 replied:
yeah, we build to a static bundle. but i assume the favicon is provided separately, cuz it definitely works fine currently right?

MrFlashAccount replied:
Yea, but I'm currently talking about dev mode, we can put favicon in ./public folder and it'll be served automatically

somebody1234 replied:
oh, yes vite is used for production builds, which is why it's not in `/public`. there is an esbuild script in dashboard but to my knowledge it is no longer used

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/components/styled/Button.tsx:
```diff
@@ -0,0 +1,72 @@
+/** @file A styled button. */
```
Can we mark this button as deprecated in favour of `aria/Button` ?

somebody1234 replied:
?

somebody1234 replied:
ah, maybe. i might have to look into it, or you could do it in another PR

MrFlashAccount reviewed (commented):
There is still a lot review, but I'd like to leave some comments upfront:

MrFlashAccount commented:
@somebody1234 I hope in the next PRs we will separate refactorings and implementation of actual behavior into different PRs. It's really hard to do code review of unrelated changes(renames, refactorings, etc)

somebody1234 commented:
this PR _is_ supposed to be a refactor though. and while a lot of files were changed, it should be clear enough which ones are minor changes (mostly for consistency) and which are the main part of the refactor. 

imo the *biggest* issue is the addition of `FocusArea` which breaks diffs - although ignoring whitespace changes mostly fixes that:
<https://github.com/enso-org/enso/pull/9499/files?w=1>

apart from that, the biggest change should be the splitting of authentication pages into individual components. this was previously *required* pre-`FocusArea` otherwise it would break because the execution order of the `ref`s was broken - i just figured it doesn't hurt to keep it

<blockquote><img src="https://avatars.githubusercontent.com/u/4046547?s=400&v=4" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso/pull/9499">Keyboard navigation between components by somebody1234 · Pull Request #9499 · enso-org/enso</a></strong></div><div>Pull Request Description

Close enso-org/cloud-v2#982

Add keyboard navigation via arrows between different components

This is achieved by a Navigator2D class which keeps track of the closest adja...</div></blockquote>

MrFlashAccount commented:
⛔️ QA did not pass

1. We have a duplicate focus on both items in the assets table and the columns head: 
    ![CleanShot 2024-04-02 at 11 07 43@2x](https://github.com/enso-org/enso/assets/61194245/2da56d50-ac6b-4cdd-8058-b3cbcf917da7)
  How to reproduce: use top-bottom arrows to navigate between items, then click the right arrow.
2. When I click left on a collapsed folder I expect the focus to jump up to the parent folder, but the focus jumps to the sidebar:
https://github.com/enso-org/enso/assets/61194245/98131539-051d-4033-bdad-628583fd5319
3. I can't scroll the table using the arrow keys.
4. A delete label icon doesn't appear when I focus on it:
![CleanShot 2024-04-02 at 11 20 33@2x](https://github.com/enso-org/enso/assets/61194245/ed456d77-2789-4d27-b862-16e77c14cf2c)
5. When I click the "right arrow" in the sidebar I expect to focus on a table item
https://github.com/enso-org/enso/assets/61194245/d3903bcd-c952-46a5-a9fe-63080b710ac3
6. When I click the "bottom arrow" in the search bar, I expect to navigate within the search bar dropdown but I get jump to the table. Jumping outside of the search bar should only happen when I click "tab"
7. When I click "tab" in search bar, it navigates between suggestions, also it applies them, but doesn't show these suggestions as applied: 

https://github.com/enso-org/enso/assets/61194245/fa61f8c3-6468-4112-ac94-91fe21a35a55
How it should work: 

https://github.com/enso-org/enso/assets/61194245/3e3fda68-c003-4077-8ad3-cf37cb747179

8. I can't share a project with someone using keyboard navigation.
9. In settings, in the keyboard shortcuts table, at the bottom of the list, click a few times the "up arrow", and focus jumps to the search bar. 

https://github.com/enso-org/enso/assets/61194245/a9022f0a-29ac-472a-afff-af327cb8eeea

10. I can't focus filter bubbles inside the search bar: 
![CleanShot 2024-04-02 at 11 37 31@2x](https://github.com/enso-org/enso/assets/61194245/2ef8b6b6-dc7d-4b7c-bdd8-41c5950bc58e)

11. I think using "shift" key to toggle bw "include" and "exclude" filter options in the search bar becomes obsolete, bc shift+tab should move focus to the prev item, thus making these filters unreachable.
12. I can't add labels using keyboard navigation

I think that using arrow keys to navigate between adjacent elements(that are not within a group, e.g. menu items, combobox items, button groups, radio groups, etc) isn't a good idea, because it violates the A11Y guidelines and often leads to unexpected behavior.
For example:
![CleanShot 2024-04-02 at 11 44 45@2x](https://github.com/enso-org/enso/assets/61194245/e869e418-3b22-48b6-be70-c5a411e4bb3a)
Here, on the screenshot I selected different areas. Inside these areas, I can use arrow keys to navigate between items, but I need to use tab navigation to jump between areas. This is what people are used to.
Another example is Google Drive:

https://github.com/enso-org/enso/assets/61194245/772243f6-7f09-4248-be9e-6f9a09849275

Here we see that we can navigate inside areas using arrows, but to navigate between areas, we use tab navigation.

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/pages/subscribe/Subscribe.tsx:
```diff
@@ -16,6 +16,7 @@ import * as backendProvider from '#/providers/BackendProvider'
 import * as textProvider from '#/providers/TextProvider'
 
 import Modal from '#/components/Modal'
+import UnstyledButton from '#/components/styled/UnstyledButton'
```
Not sure it's a good idea to put `UnstyledButton` under the `styled` folder

somebody1234 replied:
true... the long-term intention was to slowly turn all components into `styled` components and then move `styled/*` into the root. but to be fair it doesn't hurt to *not* put this file in the wrong place...

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/pages/authentication/SetUsername.tsx:
```diff
@@ -10,8 +10,11 @@ import * as backendProvider from '#/providers/BackendProvider'
 import * as textProvider from '#/providers/TextProvider'
 
 import Input from '#/components/Input'
+import AuthenticationPage from '#/components/styled/AuthenticationPage'
```
Why do we put parts of the page that are not supposed to be used somewhere else into a common `#/components` folder, instead of putting them into a local folder near the `authentication` folder?

somebody1234 replied:
good point. i guess the real issue is that the `layouts/`folder isn't really the best name for what it's used for

(edit: `pages/` too. maybe we should switch to something like feature sliced design at some point in the future.)

somebody1234 commented:
11. it's useful when using mouse. i think shift is still the most intuitive modifier, but you're welcome to suggest a better one.
12. i can't find where it violates a11y guidelines - neither in [the wcag guidelines](https://www.w3.org/WAI/WCAG21/Techniques/general/G202) or anywhere else
  - it's true that it's *unusual*, but i'm not sure it's *worse*. and i can see this being useful - i do think it's a pain relying on tab order because tab order typically cycles though *every focusable element*. and so you end up holding down `tab` for extended periods of time waiting until you're near the correct place in the document.
  - note that this shouldn't affect a11y for screenreaders and the like because afaict they rely solely on tab order. although we might want to add affordances sometime, i'm not sure it's a particularly high priority given that afaict the GUI (the graph editor) is not yet accessible

nice catches though, i'll definitely have to investigate them

MrFlashAccount commented on app/ide-desktop/lib/dashboard/src/modals/UpsertSecretModal.tsx:
```diff
@@ -49,56 +54,79 @@ export default function UpsertSecretModal(props: UpsertSecretModalProps) {
+                    autoFocus
+                    disabled={!isNameEditable}
+                    placeholder={getText('secretNamePlaceholder')}
+                    className="focus-child text grow rounded-full border border-primary/10 bg-transparent px-input-x selectable enabled:active"
```
I'm not sure it's a good idea to use implicit contract to make an element focusable. I think we should hide this behind a hook and use the [mergeProps](https://react-spectrum.adobe.com/react-aria/mergeProps.html) function to apply this behavior.

somebody1234 committed:
Address review (move components)

somebody1234 committed:
WIP: Refactor `focusChild` out to a hook

The rest cannot be converted yet as focus children need to be in separate components

somebody1234 committed:
WIP: Address QA

somebody1234 committed:
Simplify `FocusArea` and `FocusRoot`

somebody1234 committed:
WIP: Address QA

somebody1234 commented:
status update (will be updated periodically):
1. assets table -> columns works. columns -> assets table still needs more investigating
2. done locally
3. semi-intentional. scrolling *right* works... i suppose i should fix scrolling *left* soonish.
4. done locally
5. fixed by unrelated changes, i think?
6. fixed
7. can't repro (neither before nor after the switch with `Tab` <-> `Arrow`)
8. done locally. (for "add label" plus button too)
9. can't repro?
10. semi intentional. i think this one needs to be thought out properly so might be out of scope for this PR.
11. a non-issue as mentioned above. however it's true that negative filters are currently unreachable using keyboard navigation, although it's worth noting (as in 10) that it's probably a better idea to properly figure out search bar navigation rather than applying a bandaid in this PR
12. makes sense but i'm not sure what solution we want to implement. should we remove arrow key navigation completely? hide it behind a modifier? hide it behind a setting?

somebody1234 committed:
Switch arrows/tab behavior for asset search bar; fix tab order of user menu

somebody1234 committed:
Make Asset Panel clickable again

somebody1234 committed:
Make labels tab-focusable

They still cannot be *selected* using keyboard, though (neither with Space nor with Enter)

somebody1234 committed:
Improvements to tab order

somebody1234 committed:
Make chat unfocusable when it is hidden

somebody1234 committed:
Merge branch 'develop' into wip/sb/keyboard-navigate-between-sections

somebody1234 committed:
Fix React warning

somebody1234 committed:
Fix keyboard navigation for date picker

somebody1234 commented:
@MrFlashAccount should be ready for re-QA i think?

a couple minor things:
- still not too sure what to do with the arrow key navigation. see reply to issue 12 above.
- i think we might want to just merge this in soonish even with a couple of issues. it's certainly far from ideal but it's already pretty huge (+ it's taken quite a lot of time) - anything you think makes sense to be fixed afterwards can be added as an issue and fixed later maybe?
  - of course, the major issues should still be fixed here

- `focus-child` is still written out explicitly in a bunch of places. it's certainly not ideal, but imo refactors belong in another PR as well?

MrFlashAccount commented:
@somebody1234 Will do QA today!

> i think we might want to just merge this in soonish even with a couple of issues. it's certainly far from ideal but it's already pretty huge

100%, I'm sure we'll fix most of the issues iteratively with leverage from react-aria

> focus-child is still written out explicitly in a bunch of places. it's certainly not ideal, but imo refactors belong in another PR as well?

Agree.

---

According to the arrow keys navigation:
I think it's fine to go with it for now, but I'm not sure it's a feature worth supporting (and testing!) in the future. Also, if the arrow key navigation breaks the patterns people are used to, we should prefer these patterns over the navigation(IMO).

MrFlashAccount commented:
✅ QA done.

There are a few minor issues, but we'll eventually fix them.

somebody1234 committed:
Prettier

somebody1234 committed:
Fix E2E tests

somebody1234 committed:
Merge branch 'develop' into wip/sb/keyboard-navigate-between-sections

somebody1234 committed:
Merge branch 'develop' into wip/sb/keyboard-navigate-between-sections

somebody1234 committed:
Fix secret modal
